### DISCUSSION TITLE
[Proposal]: Nested Generics

### DISCUSSION THREAD

**Discussion by martindrlik** [dotnet/csharplang#4023] [Language Ideas] [answered]
# Nested Generics

* [x] Proposed
* [ ] Prototype: Not Started
* [ ] Implementation: Not Started
* [ ] Specification: Not Started

## Summary
[summary]: #summary

Type figured out from the nested generics without directly specifying.

## Motivation
[motivation]: #motivation

No need to specify type that can be figured out from nested generics. I believe that example helps me clarify the idea. For instance MyHandler can implement generic interface `IHandler<TRequest, TResponse>` and in case that `TRequest` implements `IRequest<TResponse>` it might be possible to find out `TResponse` type from `IRequest`. With nested generics we could define `IHandler<TRequest<TResponse>>` and, while implementing, we can provide only request type, response type is figured out from request:
```
class MyResponse
class MyRequest : IRequest<MyResponse>
class MyHandler : IHandler<MyRequest> // response type is figured out without specifying
```

## Detailed design
[design]: #detailed-design

Detailed code example:
```
class MyHandler : IHandler<MyRequest> {
    // Note that MyResponse is figured out from MyRequest "nested generic"
    ValueTask<MyResponse> SendAsync(MyRequest request) { ... }
}

interface IHandler<TRequest<TResponse>> // "nested generics"
    where TRequest : IRequest<TResponse> {
    ValueTask<TResponse> SendAsync(TRequest request);
}

class MyResponse { ... }
class MyRequest : IRequest<MyResponse> { ... }
...
interface IRequest<TResponse> { ... }
```

## Drawbacks
[drawbacks]: #drawbacks

## Alternatives
[alternatives]: #alternatives

Currently I need to explicitly specify all types even if some would be possible to figure out:
```
class MyResponse
class MyRequest : IRequest<MyResponse>
class Handler : IHandler<MyRequest, MyResponse> // MyResponse already mentioned for MyRequest
```

## Unresolved questions
[unresolved]: #unresolved-questions

**Comment by Frassle** [comment, marked as answer]
I think this is a duplicate of https://github.com/dotnet/csharplang/issues/339
